### DISCUSSION TITLE
Erode "kernel" settings in Double Exposure tool

### DISCUSSION THREAD

**Discussion by futrtrubl** [sn4k3/UVtools#367] [Ideas]
Can we get the same kernel settings as in the elephant's foot calibrator?

**Comment by sn4k3** [comment]
I can, but i think it's not worth it.
Are you using kernels other than default?

Also does the Double Exposure tool produce good results for you?

**Reply by futrtrubl** [reply]
The problem is the default kernal doesn't uniformly erode borders, for instance round holes become square. Concave angles erode differently depending on their orientation. I'm using an 8 iteration erode and the effects are very noticeable. If however I make a 17x17 ellipse and do one iteration it's much closer to a uniform reduction.
As to producing good results, I find it does. My only issues with it (any particular setting will erode too much on convex surfaces and too little on concave) would affect any method unless there is some compensation for the concavity/convexity of the nearby surface, which sounds too complex. When I did the elephant's foot calibration none of the wall dimming ones worked well for me. I will try again though with much greater dimming.

**Reply by futrtrubl** [reply]
Maybe a better way to erode would be that instead of multiple iterations of a 3x3 kernel, which would either overerode or undererode on diagonals for square and cross shaped respectively, choose the erode distance d and then make a 2d+1 elliptical kernel and do just one iteration.

**Reply by sn4k3** [reply]
Ok i will have a look at it

**Comment by sn4k3** [comment]
The 2d+1 elliptical kernel works best, but the biggest the kernel the slowest is. Still should not be a problem since this used for bottoms only. More guys with same problem: https://stackoverflow.com/questions/31025368/erode-is-too-slow-opencv/31044305
Here the sample with 40 iterations:

Original:
![image](https://user-images.githubusercontent.com/113281/145768565-937845c6-9346-45f2-b0cd-9fbfc403af4e.png)

3x3 rectangle:
![image](https://user-images.githubusercontent.com/113281/145768443-ade81f00-b95e-4c4c-b8f1-affd41c1cda1.png)

dynamic kernel:
![image](https://user-images.githubusercontent.com/113281/145768518-b5e977de-9d50-40ef-9f9a-ef0d3089c595.png)

**Reply by futrtrubl** [reply]
I suppose an option for a dynamic kernel isn't very necessary if the user can select a large kernel manually. I assume Pixel Arithmetic also uses erode for wall detection for eg. elephant's foot compensation? Exposing the setting everywhere it is used may be a problem. Perhaps a button where it is used to a separate kernel setting popup would allow it to be accessed without cluttering the UI?
The only place a dynamic kernel might still be useful would be for the Elephant's Foot Calibration where many different erodes would be done.

**Reply by sn4k3** [reply]
I already have the interface done by adding 1 option, it doesnt hurt

![image](https://user-images.githubusercontent.com/113281/145864571-b7194ece-af20-48a0-8d83-41e8f2ad4f52.png)

![image](https://user-images.githubusercontent.com/113281/145864598-35981149-47c1-4084-80e8-7a47ecceb0ca.png)

I can  move that checkbox inside the Kernel expander, but users may skip it all the time, so i'm not sure where it should go better

![UVtools_2021-12-13_18-05-31](https://user-images.githubusercontent.com/113281/145864963-96a6a716-1e7b-4c91-849f-e3d8a2a56fac.png)


> I assume Pixel Arithmetic also uses erode for wall detection for eg. elephant's foot compensation?

This is correct

**Reply by futrtrubl** [reply]
Looks great!
